aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -3,139 +3,202 @@ title: Dependency injection in ASP.NET Core
+An instance of the `MyDependency` class can be created to make the `WriteMessage` method available to a class. The `MyDependency` class is a dependency of the `IndexModel` class:
 
-### Constructor injection behavior
-
-Constructor injection requires that the constructor in question be *public*. Otherwise, your app will throw an `InvalidOperationException`:
+```csharp
+public class IndexModel : PageModel
+{
+    MyDependency _dependency = new MyDependency();
 
-> A suitable constructor for type 'YourType' couldn't be located. Ensure the type is concrete and services are registered for all parameters of a public constructor.
+    public async Task OnGetAsync()
+    {
+        await _myDependency.WriteMessage(
+            "IndexModel.OnGetAsync created this message.");
+    }
+}
+```
 
-Constructor injection requires that only one applicable constructor exist. Constructor overloads are supported, but only one overload can exist whose arguments can all be fulfilled by dependency injection. If more than one exists, your app will throw an `InvalidOperationException`:
+::: moniker-end
 
-> Multiple constructors accepting all given argument types have been found in type 'YourType'. There should only be one applicable constructor.
+::: moniker range="<= aspnetcore-2.0"
 
-Constructors can accept arguments that are not provided by dependency injection, but these must support default values. For example:
+An instance of the `MyDependency` class can be created to make the `WriteMessage` method available to a class. The `MyDependency` class is a dependency of the `HomeController` class:
 
 ```csharp
-// throws InvalidOperationException: Unable to resolve service for type 'System.String'...
-public CharactersController(ICharacterRepository characterRepository, string title)
+public class HomeController : Controller
 {
-    _characterRepository = characterRepository;
-    _title = title;
-}
+    MyDependency _dependency = new MyDependency();
 
-// runs without error
-public CharactersController(ICharacterRepository characterRepository, string title = "Characters")
-{
-    _characterRepository = characterRepository;
-    _title = title;
+    public async Task<IActionResult> Index()
+    {
+        await _myDependency.WriteMessage(
+            "HomeController.Index created this message.");
+
+        return View();
+    }
 }
 ```
 
-## Using framework-provided services
+::: moniker-end
 
-The `ConfigureServices` method in the `Startup` class is responsible for defining the services the application will use, including platform features like Entity Framework Core and ASP.NET Core MVC. Initially, the `IServiceCollection` provided to `ConfigureServices` has the following services defined (depending on [how the host was configured](xref:fundamentals/host/index)):
+The class creates and directly depends on the `MyDependency` instance. However, it's a bad practice to hardcode the dependency in this way for the following reasons:
```

Comment:
~However, it's a bad practice to hardcode the dependency in this way for the following reasons:~

We should avoid bad. Hardcode is jargon and should be avoided. Maybe something like

Code dependencies (such as the previous example) are problematic and should be avoided for the following reasons: